src/python/grpcio/grpc/__init__.py

```diff
@@ -27,4 +27,716 @@
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def code(self):
+    """Accesses the status code emitted by the service-side of the RPC.
+
+    This method blocks until the value is available.
+
+    Returns:
+      The integer status code of the RPC.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def details(self):
+    """Accesses the details value emitted by the service-side of the RPC.
+
+    This method blocks until the value is available.
+
+    Returns:
+      The bytes of the details of the RPC.
+    """
+    raise NotImplementedError()
+
+
+class ServicerContext(six.with_metaclass(abc.ABCMeta, RpcContext)):
+  """A context object passed to method implementations."""
+
+  @abc.abstractmethod
+  def invocation_metadata(self):
+    """Accesses the metadata from the invocation-side of the RPC.
+
+    Returns:
+      The invocation metadata object as a sequence of pairs of bytes.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def initial_metadata(self, initial_metadata):
+    """Accepts the initial metadata value of the RPC.
+
+    This method need not be called by method implementations if they have no
+    service-side initial metadata to transmit.
+
+    Args:
+      initial_metadata: The initial metadata of the RPC as a sequence of pairs
+        of bytes.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def terminal_metadata(self, terminal_metadata):
+    """Accepts the terminal metadata value of the RPC.
+
+    This method need not be called by method implementations if they have no
+    service-side terminal metadata to transmit.
+
+    Args:
+      terminal_metadata: The terminal metadata of the RPC as a sequence of pairs
+        of bytes.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def code(self, code):
+    """Accepts the status code of the RPC.
+
+    This method need not be called by method implementations if they wish the
+    gRPC runtime to determine the status code of the RPC.
+
+    Args:
+      code: The integer status code of the RPC to be transmitted to the
+        invocation side of the RPC.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def details(self, details):
+    """Accepts the service-side details of the RPC.
+
+    This method need not be called by method implementations if they have no
+    details to transmit.
+
+    Args:
+      details: The details bytes of the RPC to be transmitted to
+        the invocation side of the RPC.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def peer(self):
+    """Identifies the peer that invoked the RPC being serviced.
+
+    Returns:
+      A string identifying the peer that invoked the RPC being serviced.
+    """
+    raise NotImplementedError()
+
+
+class UnaryUnaryMultiCallable(six.with_metaclass(abc.ABCMeta)):
+  """Affords invoking a unary-unary RPC."""
```

Comment:
> **I think we've heard quite loudly from our users that the grpc Python package is where they expect every developer-user-facing code element to be found.**

Was it that they wanted it in `grpc`, or was it that they didn't want to dive through five layers of packages to find user-facing visible elements? Even if such users suggested the bundling of everything into the top-level namespace, do we know if those users picking *that particular organization* wasn't out of exasperation? There's a tendency to swing pendulums far from equilibrium when starting far from equilibrium in the opposite direction, and many, many well-used pre-existing not-complained-about-much libraries in Python (which I interpret as the equilibrium position of the Python community) follow neither the excessive partitioning of our first attempt nor the totally flat namespace in this approach.

> [...] spread code elements out reasonably into files like `_future.py`, `_multi_callables.py`, `_context.py`, and then alias them in `grpc/__init__.py` [...]

I'm not okay with this. I know that I'd mentioned being okay with something like it earlier in meat-space but IIRC that was in the context of bandaging a pre-existing organization (e.g. if we wanted to switch from a flat organization to a tree or vice-versa), and this is a *fresh* organization. Also, private modules for interfaces leave me feeling... knotted.

> Please insist on breaking it up if you feel strongly, otherwise have I persuaded you at all?

I'm not persuaded, but breaking up the organization looks like it'd be yet another can of worms. The totally flat namespace is significantly better than what we had before and I consent to it. Just, seriously consider adding comment section headers.